vmdk_plugin/vmdkops/esx_vmdkcmd.go

```diff
@@ -85,19 +86,30 @@ func (vmdkCmd EsxVmdkCmd) Run(cmd string, name string, opts map[string]string) (
 	defer C.free(unsafe.Pointer(ans))
 
 	// connect, send command, get reply, disconnect - all in one shot
-	_, err = C.Vmci_GetReply(C.int(EsxPort), cmdS, beS, ans)
-
-	if err != nil {
-		var errno syscall.Errno
-		errno = err.(syscall.Errno)
-		msg := fmt.Sprintf("'%s' failed: %v (errno=%d).", cmd, err, int(errno))
-		if errno == syscall.ECONNRESET {
-			msg += " Hit communication issue with ESX (vmci or ESX service)\n"
-			msg += " Please refer to the FAQ https://github.com/vmware/docker-volume-vsphere/wiki#faq"
+	retryCount := 10
```

Comment:
Not really, the plugin server returns an error that the disk is already
attached on a second attach (without detaching the volume). Now, do we
really want the esx_vmdkcmd.go handling server errors? It knows nothing of
volumes.

a. A retry can cause the server to return an error.

b. If the error is passed up then plugin.go would assume that the disk
attach failed and will return to the caller (in plugin.go:mountVolume()).

c. It will not try to mount the volume on the client.

To be idempotent, the server must return OK status vs. error. Or the client
must fetch status and figure where the volume is mounted.

Docker 1.12 allows returning plugin specific status from server and that
can be leveraged to figure if the volume is attached to the client.

On Thu, Jun 30, 2016 at 1:31 PM, Mark Sterin <notifications@github.com>
wrote:

> In vmdk_plugin/vmdkops/esx_vmdkcmd.go
> <https://github.com/vmware/docker-volume-vsphere/pull/499#discussion_r69087644>
> :
>
> > @@ -85,19 +86,30 @@ func (vmdkCmd EsxVmdkCmd) Run(cmd string, name string, opts map[string]string) (
> >  	defer C.free(unsafe.Pointer(ans))
> >
> >  	// connect, send command, get reply, disconnect - all in one shot
> > -	_, err = C.Vmci_GetReply(C.int(EsxPort), cmdS, beS, ans)
> > -
> > -	if err != nil {
> > -		var errno syscall.Errno
> > -		errno = err.(syscall.Errno)
> > -		msg := fmt.Sprintf("'%s' failed: %v (errno=%d).", cmd, err, int(errno))
> > -		if errno == syscall.ECONNRESET {
> > -			msg += " Hit communication issue with ESX (vmci or ESX service)\n"
> > -			msg += " Please refer to the FAQ https://github.com/vmware/docker-volume-vsphere/wiki#faq"
> > +	retryCount := 10
>
> Actually, that's exactly how the current code works (If the volume is
> attached second attach is a no-op).
> For detach it's different (for historical reasons) but could be easily
> changed with no side effects - just open a bug and assign to me, or fix
> detach_disk:713 (return err(msg)) to return None , and adjust the log msg
> to "Info"
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/vmware/docker-volume-vsphere/pull/499/files/23e50189635c854f88aaf3f885bcc0188045c2ed#r69087644>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/APHseO0sro3sPRgTrS0sE3gPgcQ6sjiJks5qQ3e_gaJpZM4JBwl3>
> .
>